saleor/graphql/product/types.py

```diff
@@ -16,31 +16,55 @@
 
 
 def resolve_attribute_list(attributes):
-    attribute_list = []
-    if attributes:
-        product_attributes = dict(
-            models.ProductAttribute.objects.values_list('id', 'slug'))
-        attribute_values = dict(
-            models.AttributeChoiceValue.objects.values_list('id', 'slug'))
-        for k, v in attributes.items():
-            value = None
-            name = product_attributes.get(int(k))
-            if v:
-                value = attribute_values.get(int(v))
-            attribute_list.append(
-                SelectedAttribute(name=name, value=value))
+    keys = list(attributes.keys())
+    values = list(attributes.values())
+    product_attributes = list(
+        models.ProductAttribute.objects.filter(pk__in=keys))
+    product_values = list(
+        models.AttributeChoiceValue.objects.filter(pk__in=values))
+
+    attribute_list = [SelectedAttribute(
+        attribute=attr, attribute_value=attr_value)
+        for k, v in attributes.items()
+        for attr in product_attributes if attr.pk == int(k)
+        for attr_value in product_values if attr_value.pk == int(v)]
```

Comment:
This would be significantly shorter and possibly more readable if `product_attributes` and `product_values` were rewritten as maps (`{x.pk: x for x in ...}`).

Reply:
I think `product.utils.attributes.get_product_attributes_data` does exactly the thing. We shuold try to reuse it here.

Reply:
I refactored it the way @patrys suggested; the problem with `get_product_attributes_data` is that it is designed to work on `product.product_type.product_attributes`, while in API it is not always a case - sometimes `variant_attributes` are needed. Few lines of code here seems to be a better option than trying to improve `get_product_attributes_data` right now - it could be a subject of separate issue possibly.